device/transport/mqtt/src/mqtt.ts

```diff
@@ -70,6 +71,183 @@ export class Mqtt extends EventEmitter implements Client.Transport, StableConnec
+          disconnect: (disconnectCallback) => {
+            this._fsm.transition('disconnecting', disconnectCallback);
+          },
+          sendEvent: (topic, payload, options, sendEventCallback) => {
+            this._mqtt.publish(topic, payload, options, (err, result) => {
+              if (err) {
+                /*Codes_SRS_NODE_DEVICE_MQTT_16_027: [The `sendEvent` method shall call its callback with an `Error` that has been translated using the `translateError` method if the `MqttBase` object fails to publish the message.]*/
+                sendEventCallback(translateError(err));
+              } else {
+                sendEventCallback(null, new results.MessageEnqueued(result));
+              }
+            });
+          },
+          sendTwinRequest: (topic, body, callback) => {
+            /* Codes_SRS_NODE_DEVICE_MQTT_18_001: [** The `sendTwinRequest` method shall call the publish method on `MqttTransport`. **]** */
+            /* Codes_SRS_NODE_DEVICE_MQTT_18_015: [** The `sendTwinRequest` shall publish the request with QOS=0, DUP=0, and Retain=0 **]** */
+            this._mqtt.publish(topic, body.toString(), { qos: 0, retain: false }, (err, puback) => {
+              if (err) {
+                /* Codes_SRS_NODE_DEVICE_MQTT_18_016: [** If an error occurs in the `sendTwinRequest` method, the `done` callback shall be called with the error as the first parameter. **]** */
+                /* Codes_SRS_NODE_DEVICE_MQTT_18_024: [** If an error occurs, the `sendTwinRequest` shall use the MQTT `translateError` module to convert the mqtt-specific error to a transport agnostic error before passing it into the `done` callback. **]** */
+                callback(translateError(err));
+              } else {
+                /* Codes_SRS_NODE_DEVICE_MQTT_18_004: [** If a `done` callback is passed as an argument, The `sendTwinRequest` method shall call `done` after the body has been published. **]** */
+                /* Codes_SRS_NODE_DEVICE_MQTT_18_017: [** If the `sendTwinRequest` method is successful, the first parameter to the `done` callback shall be null and the second parameter shall be a MessageEnqueued object. **]** */
+                callback(null, new results.MessageEnqueued(puback));
+              }
+            });
+          },
+          updateSharedAccessSignature: (sharedAccessSignature, callback) => {
+            /*Codes_SRS_NODE_DEVICE_MQTT_16_028: [The `updateSharedAccessSignature` method shall call the `updateSharedAccessSignature` method on the `MqttBase` object if it is connected.]*/
+            this._mqtt.updateSharedAccessSignature(sharedAccessSignature, (err) => {
+              if (err) {
+                /*Codes_SRS_NODE_DEVICE_MQTT_16_009: [The `updateSharedAccessSignature` method shall call the `done` method with an `Error` object if `MqttBase.updateSharedAccessSignature` fails.]*/
+                this._fsm.transition('disconnected', callback, err);
```

Comment:
>this._fsm.transition('disconnected', callback, err); [](start = 16, length = 52)

This assumes that the base was disconnected.  Do we want to force the base to disconnect here to avoid inconsistent states?

Reply:
errors are emitted only when the base is disconnected so this is assumed. happy to question that pattern though.

---
In reply to: [141443815](https://github.com/Azure/azure-iot-sdk-node/pull/92#discussion_r141443815) [](ancestors = 141443815)